core/src/cache.rs

```diff
@@ -220,6 +221,65 @@ impl<E> CacheError<E> {
     }
 }
 
+/// Input data usually comes from files on the file system, but there are also
+/// lots of cases where we want to synthesize other kinds of inputs.
+///
+/// Note that a `SourcePath` does not uniquely identify a cached input:
+/// - Some functions (like [`Cache::add_file`]) add a new cached input unconditionally.
+/// - [`Cache::get_or_add_file`] will add a new cached input at the same `SourcePath` if
+///   the file on disk was updated.
+///
+/// The equality checking of `SourcePath` only affects [`Cache::replace_string`], which
+/// overwrites any previous cached input with the same `SourcePath`.
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub enum SourcePath {
+    /// A file at the given path.
+    ///
+    /// Note that this does not need to be a real file on the filesystem: it could still
+    /// be loaded from memory by, e.g, [`Cache::add_string`].
+    Path(PathBuf),
+    /// A subrange of a file at the given path.
+    ///
+    /// This is used by nls to analyze small parts of files that don't fully parse. The
+    /// original file path is preserved, because it's needed for resolving imports.
+    Snippet(PathBuf),
+    Std(StdlibModule),
+    Query,
+    ReplInput(usize),
+    ReplTypecheck,
+    ReplQuery,
+}
+
+impl SourcePath {
+    /// Return this import's path if it has one.
+    ///
+    /// The path is important for resolving relative imports. Note that different
+    /// `SourcePath`s may have the same path.
+    fn as_path(&self) -> Option<&OsStr> {
```

Comment:
I went with `TryInto` for now, because the path is needed in order to fulfill the `ImportResolver` interface. Maybe the `ImportResolver` interface could also be converted to give parents only, but I haven't dug into that yet. While checking, I did notice that there's a redundant (and possibly information-losing) FileId -> PathBuf -> FileId round trip in `ImportResolver`, so I got rid of that.